### PR TITLE
net: add UnixStream readiness and non-blocking ops

### PR DESCRIPTION
**Pull request by cssivision**
Refs: https://github.com/tokio-rs/tokio/issues/2968

### CODE REVIEW

File: tokio/src/net/unix/stream.rs [hunk 1]
```diff
@@ -43,6 +43,271 @@ impl UnixStream {
...
+    ///         let ready = stream.ready(Interest::READABLE | Interest::WRITABLE).await?;
+    ///
+    ///         if ready.is_readable() {
+    ///             // The buffer is **not** included in the async task and will only exist
+    ///             // on the stack.
+    ///             let mut data = [0; 1024];
+    ///             let n = stream.try_read(&mut data[..]).unwrap();
+    ///
+    ///             println!("GOT {:?}", &data[..n]);
+    ///         }
```
Comment (Darksonn) [inline tokio/src/net/unix/stream.rs:78]: This should check for false positives. The fact that `TcpStream`'s example does not is a mistake in the doc on `TcpStream`. I opened #3248 to track that.

File: tokio/src/net/unix/stream.rs [hunk 2]
```diff
@@ -43,6 +43,271 @@ impl UnixStream {
...
+    /// ```
+    pub async fn ready(&self, interest: Interest) -> io::Result<Ready> {
+        let event = self.io.registration().readiness(interest).await?;
+        Ok(event.ready)
+    }
+
+    /// Wait for the socket to become readable.
+    ///
+    /// This function is equivalent to `ready(Interest::READABLE)` is usually
+    /// paired with `try_read()`.
```
Comment (Darksonn) [inline tokio/src/net/unix/stream.rs:95]:
```suggestion
    /// This function is equivalent to `ready(Interest::READABLE)` and is usually
    /// paired with `try_read()`.
```

File: tokio/src/net/unix/stream.rs [hunk 3]
```diff
@@ -43,6 +43,271 @@ impl UnixStream {
...
+    ///
+    /// Usually, [`readable()`] or [`ready()`] is used with this function.
+    ///
+    /// [`readable()`]: UnixStream::readable()
+    /// [`ready()`]: UnixStream::ready()
+    ///
+    /// # Return
+    ///
+    /// If data is successfully read, `Ok(n)` is returned, where `n` is the
+    /// number of bytes read. `Ok(n)` indicates the stream's read half is closed
```
Comment (Darksonn) [inline tokio/src/net/unix/stream.rs:158]:
```suggestion
    /// number of bytes read. `Ok(0)` indicates the stream's read half is closed
```

File: tokio/src/net/unix/stream.rs [hunk 4]
```diff
@@ -43,6 +43,271 @@ impl UnixStream {
...
+    pub fn try_read(&self, buf: &mut [u8]) -> io::Result<usize> {
+        self.io
+            .registration()
+            .try_io(Interest::READABLE, || (&*self.io).read(buf))
+    }
+
+    /// Wait for the socket to become writable.
+    ///
+    /// This function is equivalent to `ready(Interest::WRITABLE)` is usually
+    /// paired with `try_write()`.
```
Comment (Darksonn) [inline tokio/src/net/unix/stream.rs:212]:
```suggestion
    /// This function is equivalent to `ready(Interest::WRITABLE)` and is usually
    /// paired with `try_write()`.
```

**Review by Darksonn — COMMENTED** [review]
There are some minor doc things, which I assume originate from the `TcpStream` PR.

**Comment by Darksonn** [comment]
Can you add `poll_read_ready` and `poll_write_ready` too?

**Comment by cssivision** [comment]
> Can you add `poll_read_ready` and `poll_write_ready` too?

I'm glad to.

File: tokio/src/net/unix/stream.rs [hunk 5]
```diff
@@ -43,6 +43,314 @@ impl UnixStream {
...
+    ///
+    ///         if ready.is_writable() {
+    ///             // Try to write data, this may still fail with `WouldBlock`
+    ///             // if the readiness event is a false positive.
+    ///             match stream.try_write(b"hello world") {
+    ///                 Ok(n) => {
+    ///                     println!("write {} bytes", n);
+    ///                 }
+    ///                 Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+    ///                     continue
```
Comment (Darksonn) [inline tokio/src/net/unix/stream.rs:98]: Just to be consistent with the read case.
```suggestion
    ///                     continue;
```

**Review by Darksonn — APPROVED** [review]
Thanks!

**Pull request merged**